lte/gateway/c/session_manager/test/test_local_enforcer.cpp

```diff
@@ -2230,6 +2243,109 @@ TEST_F(LocalEnforcerTest, test_invalid_apn_parsing) {
       session_map, epoch, [](Status status, SetupFlowsResult resp) {});
 }
 
+TEST_F(LocalEnforcerTest, test_final_unit_activation_and_canceling) {
+  CreateSessionResponse response;
+  create_credit_update_response(
+      IMSI1, SESSION_ID_1, 1, 1024, ChargingCredit_FinalAction_RESTRICT_ACCESS,
+      "", "rule1", response.mutable_credits()->Add());
+
+  auto dynamic_rule = response.mutable_dynamic_rules()->Add();
+  auto policy_rule  = dynamic_rule->mutable_policy_rule();
+  policy_rule->set_id("rule2");
+  policy_rule->set_rating_group(1);
+  policy_rule->set_tracking_type(PolicyRule::ONLY_OCS);
+  auto static_rule = response.mutable_static_rules()->Add();
+  static_rule->set_rule_id("rule1");
+  static_rule = response.mutable_static_rules()->Add();
+  static_rule->set_rule_id("rule3");
+
+  insert_static_rule(1, "", "rule1");
+  insert_static_rule(1, "", "rule3");
+
+  // The activation for the static rules (rule1,rule3) and dynamic rule (rule2)
+  EXPECT_CALL(
+      *pipelined_client, activate_flows_for_rules(
+      testing::_, testing::_, testing::_, CheckCount(2),
+      CheckCount(1), testing::_))
+      .Times(1)
+      .WillOnce(testing::Return(true));
+
+  EXPECT_CALL(
+      *pipelined_client, update_ipfix_flow(
+      testing::_, testing::_, testing::_, testing::_,
+      testing::_, testing::_))
+      .Times(1)
+      .WillOnce(testing::Return(true));
+
+  local_enforcer->init_session_credit(
+      session_map, IMSI1, SESSION_ID_1, test_cfg_, response);
+
+  // Insert record and aggreagate thems
+  RuleRecordTable table;
+  auto& ip = test_cfg_.common_context.ue_ipv4();
+  auto record_list = table.mutable_records();
+  create_rule_record(IMSI1, ip, "rule1", 1024, 2048, record_list->Add());
```

Comment:
You'll probably have to add the IP address here: https://github.com/magma/magma/pull/2742